### *"Courage is NOT having the strength to go on; it is going on when you don't have the strength."*
*-Theodore Roosevelt*

So be courageous!  The end of the semester is in sight!

----

YOUR NAME HERE: SPENCER DAUGHERTY

----

# HW 8: Group By and Having and Join

In [4]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import sqlite3 as sql



## Review

Based on what you have learned, here's the general form of a select statement.

```SELECT column_name(s)```

```FROM table_name```

```WHERE condition```

```GROUP BY column_name(s)```

```HAVING condition```

```ORDER BY column_name(s)```


Now remember:  

The ```HAVING``` clause is like a WHERE clause but is used when the selection condition is based on aggregate funtion of a column, such as ```COUNT(name) > 10``` or ```MAX(Rain) < 2.0```.

In this homework, we will review the select statement, focusing on the GROUP BY and HAVING functionality as well as the JOIN functionality.  To make sure you are understanding when to use the capabilities and when to not, some problems will not require these techniques, so think hard before assuming you need one of these keywords.

----

Let's bring in the weather data.

Recall that the table contains about 62 years worth of temperature, rainfall and sunshine data for the city of London.

There is a record/row for each year/month from Jan 1957 to April 2019.  

* Tmax:  The max temp for the month (celsius)
* Tmin:  The min temp for the month (celsius)
* Rain: the rainfall is the total amout of rain for the month in millimeters
* Sun : the total number of hours sunshine for the month.

In [6]:
#Read in a csv with weather data
weather = pd.read_csv('weather.csv')
print(weather)
#conn = sql.connect('weather.db')
#weather.to_sql('weather', conn)

     Year  Month  Tmax  Tmin  Rain    Sun
0    1957      1   8.7   2.7  39.5   53.0
1    1957      2   9.0   2.9  69.8   64.9
2    1957      3  13.9   5.7  25.4   96.7
3    1957      4  14.2   5.2   5.7  169.6
4    1957      5  16.2   6.5  21.3  195.0
..    ...    ...   ...   ...   ...    ...
743  2018     12  10.7   5.2  60.6   40.3
744  2019      1   7.6   2.0  33.2   56.4
745  2019      2  12.4   3.3  34.2  120.2
746  2019      3  13.1   5.8  49.6  119.0
747  2019      4  15.8   5.7  12.8  170.1

[748 rows x 6 columns]


### Don't forget to use your SQL Lab notes! 


----

### Question 1.

Except for 2019, there are 12 rows per year in the weather data beacuse we have monthly weather measurements.  

Find the average rainfall per year, except for 2019.

So your final answer will be as follows.

| Year | AVG(Rain) |
|:----:|:---------:|
| 1957 | 46.466667 |
| 1958 | 61.641667 |
| 1959 | 40.608333 |
| 1960 | 61.283333 |
| 1961 | 46.566667 |
|  ... |    ...    |
| 2014 | 72.000000 |
| 2015 | 46.833333 |
| 2016 | 49.216667 |
| 2017 | 47.683333 |
| 2018 | 48.333333 |

In [3]:
sql_statement = """Select  Year, avg(Rain) 
                    from weather
                    where year not in (2019)
                    group by year"""

weather_results = pd.read_sql_query(sql_statement, conn)
weather_results

Year  avg(Rain)
0   1957  46.466667
1   1958  61.641667
2   1959  40.608333
3   1960  61.283333
4   1961  46.566667
..   ...        ...
57  2014  72.000000
58  2015  46.833333
59  2016  49.216667
60  2017  47.683333
61  2018  48.333333

[62 rows x 2 columns]

### Question 2.

Select all the rows where the Tmin and Tmax are within 4.5 degrees celsius of each other.  Show the Year, Month, Tmin, and Tmax.

HINTS:  
* The result should be a 7 row X 4 column table.
* The first row should be Jan 1966 with a Tmin of 1.0 degree celsius and a Tmax of 5.3 degrees celsius

In [4]:
sql_statement = """Select Year, Month, Tmin, Tmax 
                    from weather 
                    where (tmax-tmin) < 4.5"""

weather_results = pd.read_sql_query(sql_statement, conn)
weather_results 

Year  Month  Tmin  Tmax
0  1966      1   1.0   5.3
1  1968     12   1.0   5.4
2  1986      2  -2.7   1.7
3  1996      1   3.1   7.2
4  2000     12   4.6   9.0
5  2002     12   5.2   9.0

### Question 3.

Like us, the summer months in London are June, July, and August.

Find the years where the average Tmax during the summer months is greater than the average overall summer Tmax.  Show the average/Max/Min Tmax for these years.

The answer will have 29 rows and is shown below.

| Year | AVG(Tmax) | MAX(Tmax) | MIN(Tmax) |
|:----:|:---------:|:---------:|:---------:|
| 1957 | 22.400000 |    23.6   |    21.1   |
| 1959 | 23.666667 |    24.7   |    22.1   |
| 1970 | 22.400000 |    23.6   |    21.3   |
| 1973 | 22.333333 |    23.7   |    21.4   |
| 1975 | 23.933333 |    25.9   |    21.8   |
| 1976 | 25.733333 |    26.6   |    25.1   |
| 1983 | 24.300000 |    27.6   |    20.8   |
| 1984 | 23.300000 |    24.4   |    21.3   |
| 1989 | 24.033333 |    25.8   |    22.1   |
| 1990 | 23.400000 |    26.0   |    19.5   |
|  ... |    ...    |    ...    |    ...    |
| 2010 | 23.366667 |    25.0   |    21.6   |
| 2013 | 23.866667 |    27.0   |    20.3   |
| 2014 | 23.200000 |    25.8   |    21.7   |
| 2015 | 22.700000 |    23.7   |    22.2   |
| 2016 | 23.133333 |    24.7   |    20.7   |
| 2017 | 23.266667 |    24.0   |    22.0   |
| 2018 | 25.666667 |    28.3   |    24.2   |

In [ ]:
sql_statement = """Select year, Avg(Tmax) , Max(tmax), Min(Tmax)
                    from weather
                    where month in (6,7,8) 
                    group by year
                    having Avg(Tmax)>(Select Avg(Tmax) 
                    from weather
                    where month in (6,7,8))"""


weather_results = pd.read_sql_query(sql_statement, conn)
weather_results

### Closing the DB connection.
Let's close the DB connection as we will now work with  a different data set.

In [ ]:
conn.close()

----

### NBA Salaries

First, import the NBA Salaries from the 2015-2016 season for a few more practice problems.

In [ ]:
nba_15_16 = pd.read_csv('nba_salaries.csv') #Open the csv
conn = sql.connect('nba_more.db') #Create an nba_more database
nba_15_16.to_sql('nba_15_16', conn)  #Dump the csv into a table called nba_15_16 in the nba_more database

Now, for the first time ever, we will load in a second csv and store it as a second table in the SAME database. 

This second csv/table has some stats on the 20-21 NBA players.

Run the cell below to make this happen.

In [ ]:
nba_20_21= pd.read_csv('nba_20-21.csv')   #Open the csv
nba_20_21.to_sql('nba_20_21', conn)   #Dump it into a table called nba_20_21 in the nba_more database

If you run the cell below, you can see that our nba.db file contains two tables, namely, ```nba_15_16``` and ```nba_20_21```.   

In [ ]:
#See all the tables included
curs = conn.cursor()
curs.execute("SELECT name FROM sqlite_master WHERE type = 'table'").fetchall()

### WARNING:  

The ```FROM``` clause in all your select statements must involve one of the above table names.  You can NO LONGER say ```SELECT * FROM nba WHERE....``` This is because there is no longer a table called ```nba``` now.  Our tables are called ```nba_15_16``` and ```nba_20_21```.   

You are familiar with the first table, namely nba salaries in 2015-2016, but re-run the cell below to remnind yourself of the details. 

In [ ]:
sql_statement = """SELECT *
                   FROM nba_15_16"""

nba_results = pd.read_sql_query(sql_statement, conn)
nba_results

The data from the 20-21 table came from https://basketball.realgm.com/nba/players.  It does not include salary information but it has the following.

* jersey number
* name of player
* position of player
* height
* weight
* yos - I believe this is related to how long a player can dribble, but could not verify this.  If you know what it is, please tell me!
* predraft_team - team they were in before the draft

Let's select everything from the second table, just so that you can get a visual.  Run the cell below.

In [ ]:
sql_statement = """SELECT *
                   FROM nba_20_21"""

nba_results = pd.read_sql_query(sql_statement, conn)
nba_results

----

### Question 4.

Find how many players were on each team in the 2015-2016 season and also show the average salary of each team.  Sort the table from the team with the greatest number of players to the least.  Make the column headers match those below.

HINT: You'll get  a 30 by 3 table and the partial answer is shown below.

|                team | num_players | avg_salary |
|--------------------:|------------:|-----------:|
|   Memphis Grizzlies |          21 |   4.466497 |
|   Charlotte Hornets |          18 |   4.672355 |
|  Washington Wizards |          17 |   5.296912 |
|           Utah Jazz |          17 |   3.095993 |
|     Toronto Raptors |          17 |   4.392507 |
|        Phoenix Suns |          17 |   2.971813 |
|                 ... |         ... |        ... |
|  Los Angeles Lakers |          11 |   6.237086 |
|    Dallas Mavericks |          11 |   5.978414 |
|     Detroit Pistons |          10 |   4.221176 |
| Cleveland Cavaliers |          10 |  10.231241 |


In [ ]:
sql_statement = """Select  Team,Count(Player) as No_of_Players, Avg(Salary) as Average_Salary 
                    from nba_15_16
                    group by team
                    order by Count(Player) desc"""

nba_results = pd.read_sql_query(sql_statement, conn)
nba_results

----

### Question 5.


For each team 2015-2016, show how many players per position there were and the average salary per position.  The column titles should be shown as follows. Sort the table from the highest average salary to the least.

| position | num_players | avg_salary |
|---------:|------------:|-----------:|

HINT: You'll get  a 5 by 3 table where there are 96 shooting guards who made an average of \$3.988195 million.



In [ ]:
sql_statement = """Select Position, Count(Player) as num_players, avg(Salary) as avg_salary 
                    from nba_15_16
                    group by Position
                    order by avg(Salary) desc"""

nba_results = pd.read_sql_query(sql_statement, conn)
nba_results


Based on your analysis, what position makes the most on avereage?

*Write your answer here*

----

### Question 6.

***Using joins***, show how many players from the 2015-2016 season also played in 2021-21 season.

Hint:  The answer is between 155 and 165.

In [ ]:
sql_statement = """Select Position, Count(Player) as No_of_Players, avg(Salary) as Average_salary 
                    from nba_15_16
                    group by Position
                    order by avg(Salary) desc"""

nba_results = pd.read_sql_query(sql_statement, conn)
nba_results

### Question 7A.

***Using joins***, for each player in the 2015-2016 season, show the player's team in 2015-2016 and the player's team in 2020-2021.  Those players who were not in the NBA in the 2020-2021 season should be included.  Sort by player name and rename the columns so that they look like below.

You'll have 417 rows, since the length fo the 2015-2016 roster is 417.

|          player |          team15_16 |             team20_21 |
|----------------:|-------------------:|----------------------:|
|    Paul Millsap |      Atlanta Hawks |        Denver Nuggets |
|      Al Horford |      Atlanta Hawks | Oklahoma City Thunder |
|  Tiago Splitter |      Atlanta Hawks |                  None |
|     Jeff Teague |      Atlanta Hawks |       Milwaukee Bucks |
|     Kyle Korver |      Atlanta Hawks |                  None |
|             ... |                ... |                   ... |
|       Gary Neal | Washington Wizards |                  None |
|    DeJuan Blair | Washington Wizards |                  None |
| Kelly Oubre Jr. | Washington Wizards |                  None |
|  Garrett Temple | Washington Wizards |         Chicago Bulls |
|    Jarell Eddie | Washington Wizards |                  None |


In [ ]:
sql_statement = """Select nba_15_16.player, nba_15_16.team as team15_16, nba_20_21.team as team20_21
                    from nba_15_16
                    left join  nba_20_21
                    order by nba_15_16.player"""

nba_results = pd.read_sql_query(sql_statement, conn)
nba_results


### Question 7B.

Modify the query in 7A sp that the resulting table shows only those players who are not on the 2020-2021 roster.  ***Use the NULL keyword to do this.***

HINT:  There are 417 folks on the 2015-2016 roster and you know how many of them played in the 20-21 season from a previous question.  So how many rows should be in this table?

|      player     |      team15_16     | team20_21 |
|:---------------:|:------------------:|:---------:|
|  Tiago Splitter |    Atlanta Hawks   |    None   |
|   Kyle Korver   |    Atlanta Hawks   |    None   |
|       ...       |         ...        |    ...    |
| Kelly Oubre Jr. | Washington Wizards |    None   |
|   Jarell Eddie  | Washington Wizards |    None   |


In [ ]:
sql_statement = """Select a.player, a.team , b.Team from nba_15_16
left join  nba_20_21 b on
a.player = b.name
and  b.Team is null
order by a.player"""

nba_results = pd.read_sql_query(sql_statement, conn)
nba_results

### Question 8.

Using joins, show any players from the 2015-2016 season who are still playing but on a different team in 20-21. Be sure to show the player and both teams, renaming your columns as below.

The answer is shown below.

|     |      player     |      team15_16     |       team20_21       |
|:---:|:---------------:|:------------------:|:---------------------:|
|  0  |   Paul Millsap  |    Atlanta Hawks   |     Denver Nuggets    |
|  1  |    Al Horford   |    Atlanta Hawks   | Oklahoma City Thunder |
|  2  |   Jeff Teague   |    Atlanta Hawks   |    Milwaukee Bucks    |
|  3  |    Mike Scott   |    Atlanta Hawks   |  Philadelphia Sixers  |
|  4  |  Kent Bazemore  |    Atlanta Hawks   | Golden State Warriors |
| ... |       ...       |         ...        |          ...          |
| 129 |   Rodney Hood   |      Utah Jazz     |    Toronto Raptors    |
| 130 |    John Wall    | Washington Wizards |    Houston Rockets    |
| 131 | Markieff Morris | Washington Wizards |   Los Angeles Lakers  |
| 132 |   Jared Dudley  | Washington Wizards |   Los Angeles Lakers  |
| 133 |  Garrett Temple | Washington Wizards |     Chicago Bulls     |

In [ ]:
sql_statement = """Select a.Player, a.team as team15_16, b.team as team20_21 
                    from nba_15_16 a
                    join nba_20_21 b on
                    a.player = b.name
                    and a.team <> b.team"""

nba_results = pd.read_sql_query(sql_statement, conn)
nba_results


### Question 9.

Using joins, ***for each team***, show how many players from the 2015-2016 season also played in 2021-21 season.  Also show the average age and  average 15-16 salary of these players.  Sort by team name ascending and rename the columns so that they look like below.

You'll get 30 rows for the 30 teams.

|                team | num_in_both |   avg_age | avg_salary |
|--------------------:|------------:|----------:|-----------:|
|       Atlanta Hawks |           7 | 31.571429 |   6.724702 |
|      Boston Celtics |           5 | 28.800000 |   4.389403 |
|       Brooklyn Nets |           4 | 30.750000 |   4.374236 |
|   Charlotte Hornets |           4 | 29.500000 |   8.090966 |
|       Chicago Bulls |           6 | 29.833333 |   7.137244 |
| Cleveland Cavaliers |           4 | 31.750000 |   18.33196 |
| ...                 | ...         | ...       | ...        |


In [ ]:
sql_statement = """Select a.Player, avg(b.age), avg(a.salary) 
                    from nba_15_16 a
                    join nba_20_21 b on
                    a.player = b.name
                    group by  a.Player
                    order by b.team"""

nba_results = pd.read_sql_query(sql_statement, conn)
nba_results


### Question 10

***Using the "RIGHT" kind of join,*** :) show the name of the players in the 20-21 season, their current team, and the 2015-2016 team. 

If they dont' have a 2015-2016 team, show the phrase "NEW".

The answer is shown.

|     |           name           |       team20_21      |       team15_16       |
|:---:|:------------------------:|:--------------------:|:---------------------:|
|  0  |     Precious Achiuwa     |      Miami Heat      |          NEW          |
|  1  |       Steven Adams       | New Orleans Pelicans | Oklahoma City Thunder |
|  2  |        Bam Adebayo       |      Miami Heat      |          NEW          |
|  3  |     Ty-Shon Alexander    |     Phoenix Suns     |          NEW          |
|  4  | Nickeil Alexander-Walker | New Orleans Pelicans |          NEW          |
| ... |            ...           |          ...         |          ...          |
| 490 |       Delon Wright       |   Sacramento Kings   |    Toronto Raptors    |
| 491 |      Thaddeus Young      |     Chicago Bulls    |     Brooklyn Nets     |
| 492 |        Trae Young        |     Atlanta Hawks    |          NEW          |
| 493 |        Cody Zeller       |   Charlotte Hornets  |   Charlotte Hornets   |
| 494 |        Ivica Zubac       | Los Angeles Clippers |          NEW          |

In [ ]:
sql_statement = """Select b.name , b.team as team_20_21, isnull(a.team,'NEW') as team_15_16
                    from nba_15_16 a
                    right join nba_20_21 b
                    on a.player = b.name"""

nba_results = pd.read_sql_query(sql_statement, conn)
nba_results


### Question 11A.

For each team in 20-21, show the average weight and the min/max/average height per position.  Sort the table by the average height from high to low.  Rename your columns to have nicely formatted headers that explain what data is in the column.

There should be 9 rows.

In [ ]:
sql_statement = """Select team, position, avg(weight), min(height), max(height), average(height) 
                    from nba_20_21
                    group by team, pos
                    order by average(height) desc"""

nba_results = pd.read_sql_query(sql_statement, conn)
nba_results


### Question 11B.

Does it appear that there is a significant difference between the height and weight of any of the positions?  Explain.

*Write your answer here.*

----
Let's close the datbase connection to the nba.

In [ ]:
conn.close()

---
You're done your last HW!  Thank you for you a great semester!